F dplan 14783 extend report entries

martendemos opened the pull request:
### Ticket
DPLAN-14783



### How to review/test
Creates classes and structures to allow creation of report-entries for create, update and deletion of "documents", which means:
- elements (document-categories) (of a procedure)
- paragraph-documents (of a procedure)
- single-documents (of a procedure)
- map files (of a procedure)

On create, update and deletion of one of this listed documents, a report entry should be created.
These report-entries will not be shown in the UI yet, but needs to be created as entries in the DB.

- [ ] Create/Update tests (I would like to write the tests in a separate PR if allowed)
- [ ] Update documentation
- [ ] Add/Update data-cy attributes ([conventions](https://dplan-documentation.demos-europe.eu/development/guidelines-conventions/coding-styleguides/twig_html.html#guideline-for-naming-cypress-hooks))
- [ ] Run `yarn lint`
- [ ] Run `yarn test`
- [ ] Link all relevant tickets
- [ ] Move the tickets on the board accordingly
- [ ] Update changelog

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/Logic/Report/ParagraphReportEntryFactory.php:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * @throws JsonException
+     */
+    public function createParagraphCreateEntry(Paragraph $paragraph): ReportEntry
+    {
+        $data = $this->createData($paragraph);
+        $data['date'] = $paragraph->getCreateDate()->getTimestamp();
+        $reportEntry = $this->createParagraphReportEntry($paragraph->getProcedure()->getId(), $data);
+        $reportEntry->setCategory(ReportEntry::CATEGORY_ADD);
+
+        return $reportEntry;
+    }
+
+    /**
+     * @throws JsonException
+     */
+    public function createParagraphUpdateEntry(Paragraph $paragraph): ReportEntry
+    {
+        $data = $this->createData($paragraph);
+        $data['date'] = $paragraph->getModifyDate()->getTimestamp();
+        $reportEntry = $this->createParagraphReportEntry($paragraph->getProcedure()->getId(), $data);
+        $reportEntry->setCategory(ReportEntry::CATEGORY_UPDATE);
+
+        return $reportEntry;
+    }
+
+    /**
```
I saw in one commit that there is a Report per each entity, Could you tell me why was that? I see similar code across each of these entities for the report

martendemos replied:
I think, I don't understand the question.
What I want is, to create a report for the paragraph-entity in cases of creation, deletion or update.
May you asking why I was using three separate methods for each entity? In the meantime, I merged these three methods into one. Does this answer this question?

sierrazamudiodemos commented:
I was thinking an approach is to do it using events on create, delete, update in general on each entity, so next time a new entity needs a report, a listener can be added to it. Im curious if the event was a considered approach

mussbach commented on demosplan/DemosPlanCoreBundle/Entity/Document/Elements.php:
```diff
@@ -628,4 +645,25 @@ public function setPermission(string $permission): void
+     *
+     * @return array<string, string>
+     */
+    public function getFileInfo(): array
```
Would it be possible to use the existing FileInfo value object?

martendemos replied:
You mean as return type?
I don't know how, because at this point I do not have the path to the file and all I want is the name of the file for the report.

martendemos replied:
I could simpyl set an empty string :thinking:

martendemos replied:
I tried: 4300a22e9f13a36a3600a15c43db7e39cab1575d

mussbach commented on demosplan/DemosPlanCoreBundle/Entity/Document/SingleDocument.php:
```diff
@@ -447,4 +447,25 @@ public function getDeleteDate(): DateTime
+     *
+     * @return array<string, string>
+     */
+    public function getSingleDocumentInfo(): array
```
Would it be possible to use the existing FileInfo value object?

martendemos replied:
I tried: 4300a22e9f13a36a3600a15c43db7e39cab1575d

martendemos commented:
> I was thinking an approach is to do it using events on create, delete, update in general on each entity, so next time a new entity needs a report, a listener can be added to it. Im curious if the event was a considered approach

I'm with you. I just wanted to keep it "simple" (for me) in the first place. Anyway: I managed to use events & listener to create the report entries.

martendemos commented on demosplan/DemosPlanCoreBundle/Logic/Report/SingleDocumentReportEntryFactory.php:
```diff
@@ -0,0 +1,123 @@
+        $reportEntry->setCategory(ReportEntry::CATEGORY_DELETE);
+
+        return $reportEntry;
+    }
```
Done in multiple commits. Main one seems to be this one: 22f74c8c15e5c2c43666aa004559697794d8a9c2

sierrazamudiodemos commented:
I saw this commit: bc783f9d43794f26008e843e401757b8bfad1ec0 I was wondering if also this case is tested in other projects to make sure is always emtpy

martendemos commented on demosplan/DemosPlanCoreBundle/Logic/Report/ReportMessageConverter.php:
```diff
@@ -875,4 +833,106 @@ private function generatePeriodChangeMessage(array $message, bool $isPublic): st
+                'yes'
+            ) : $this->translator->trans('no'),
+        ]);
+    }
```
done with 78632bf35e7543ba2155ad3bb1e491e38012bba5

martendemos commented on demosplan/DemosPlanCoreBundle/EventSubscriber/ReportSubscriber.php:
```diff
@@ -54,26 +78,146 @@ public function onEditProcedure(ProcedureEditedEvent $event): void
+                /** @var SingleDocument $singleDocument */
+                $singleDocument = $entity;
+                $report = $this->singleDocumentReportEntryFactory->createSingleDocumentEntry(
+                    $singleDocument,
```
using match() in favor of switch(): 8d42f87de0687057d00a708b6517682d3b492bf8

martendemos commented on demosplan/DemosPlanCoreBundle/Logic/Document/ElementsService.php:
```diff
@@ -416,10 +419,9 @@ public function addElement(array $data): ?array
-            );
-            $this->reportService->persistAndFlushReportEntries($report);
+
+            $reportEntryEvent = new CreateReportEntryEvent($element, ReportEntry::CATEGORY_ADD);
```
Yes, there are doctrine events like postPersistEvent, postFlushEvent, etc... or even like our own ProcedureEditedEvent
I also thought about this, but in some edgecases we do want to create an entity, but not a ReportEntry.
Because of this the solution of throwing an event to indicate that a reportentry is wanted. But It also feels like a slightly misuse of events.

martendemos commented on demosplan/DemosPlanCoreBundle/Logic/Procedure/PrepareReportFromProcedureService.php:
```diff
@@ -233,14 +236,14 @@ public function createReportEntry(Procedure $sourceProcedure, Procedure $destina
-                $destinationProcedureSettings->getPlanDrawPDF()
+
+            $reportEntryEvent = new ProcedureEditedEvent(
+                $sourceProcedure->getId(),
+                ['planPDF' => $sourceProcedureSettings->getPlanPDF(), 'planDrawPDF' => $sourceProcedureSettings->getPlanDrawPDF()],
+                ['planPDF' => $destinationProcedureSettings->getPlanPDF(), 'planDrawPDF' => $destinationProcedureSettings->getPlanDrawPDF()],
+                $this->currentUser->getUser()
```
I would allow me to do it in an extra PR (the one I already have started with for the tests) later on.

martendemos commented on demosplan/DemosPlanCoreBundle/Logic/Report/ReportMessageConverter.php:
```diff
@@ -803,4 +841,162 @@ private function generatePeriodChangeMessage(array $message, bool $isPublic): st
+            $this->translator->trans('yes') : $this->translator->trans('no');
+
+        return $this->translator->trans('report.update.singleDocument', $reportEntryMessage);
+    }
```
part of 78632bf35e7543ba2155ad3bb1e491e38012bba5

sierrazamudiodemos reviewed (approved):
Thank you for addressing the concerns, I guess it is challenging to do the tests with factory oder?

martendemos commented:
> Thank you for addressing the concerns, I guess it is challenging to do the tests with factory oder?

not sure, but I will try to create all necessary factories in order to write the tests using factories
